### PR TITLE
Docker hotfix

### LINKED ISSUE #1994: Docker pull hotfix for latest
Same issue as #1984 but for v4.X.Y series.

See conversation [from here](https://github.com/alan-turing-institute/data-safe-haven/pull/1955#issuecomment-2213709499)

### PR DESCRIPTION
**Pull request by craddm**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :arrow_heading_up: Summary

Fixes DockerHub credential problem.

### :closed_umbrella: Related issues

Closes #1994

### CODE REVIEW

File: deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml [hunk 1]
```diff
@@ -89,6 +89,10 @@ write_files:
     content: |
       {{set_dns.mustache.sh}}
 
+  - path: "/opt/configuration/docker_pat.txt"
+    permissions: "0400"
+    content: {{shm.dockerPassword}}
```
Comment (JimMadge) [inline deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml:94]: I assume the default owner is `root:root`

Comment (craddm) [reply deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml:94]: Yes, and it's the `codimddaemon` user that needs the file

Comment (jemrobinson) [reply deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml:94]: Note that this gets run before any users are created so you actually *can't* give ownership to another user. 

However, given that we want the codimd user to be able to read from this, do we need to change the permissions? Or does it work as written?

File: deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml [hunk 2]
```diff
@@ -217,7 +220,8 @@ runcmd:
 
   # Deploy CodiMD using Docker
   - echo ">=== Deploying CodiMD with Docker... ===<"
-  - su codimddaemon -c "docker compose -f /opt/codimd/docker-compose.yml up -d"
+  - su codimddaemon -c "cat /opt/configuration/docker_pat.txt | docker login --username '{{shm.dockerAccount}}' --password-stdin && `
+     docker compose -f /opt/codimd/docker-compose.yml up -d"
```
Comment (JimMadge) [inline deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml]:
```suggestion
  - su codimddaemon -c "cat /opt/configuration/docker_pat.txt | docker login --username '{{shm.dockerAccount}}' --password-stdin
    && docker compose -f /opt/codimd/docker-compose.yml up -d"
```

`` ` `` isn't a continuation line in YAMl.

Comment (craddm) [reply deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml]: This is outdated

File: deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml (hunk 1 above)
Comment (craddm) [reply deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml:94]: I give ownership to the `codimddaemon` later on, which does work, but evidently getting something a little wrong, as the current error is 

```
>=== Deploying CodiMD with Docker... ===<
WARNING! Your password will be stored unencrypted in /home/codimddaemon/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

Login Succeeded
bash:  docker: command not found
```

Comment (JimMadge) [reply deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml:94]:
```yaml
  - su codimddaemon -c "cat /opt/configuration/docker_pat.txt | docker login --username '{{shm.dockerAccount}}' --password-stdin && \
      docker compose -f /opt/codimd/docker-compose.yml up -d"
```

I don't think you want the backslash here. Not sure if that is the problem though.

Is `docker compose` a command for that version of docker? If it is old enough you might still need `docker-compose`.

Comment (craddm) [reply deployment/secure_research_environment/cloud_init/cloud-init-codimd.mustache.yaml:94]: `docker compose` is correct. It was just the backslash, working now

**Comment by craddm** [comment]
Need to apply the same fix in two other places: 

- [ ] `guacamole`
- [ ] `repository-proxy`

**Comment by craddm** [comment]
Unfortunately, now hitting the exact same problem from the repository proxy servers as we were hitting earlier with requests to ubuntu update servers failing:

```
Cloud-init v. 24.1.3-0ubuntu1~20.04.5 running 'modules:config' at Tue, 09 Jul 2024 14:45:32 +0000. Up 36.76 seconds.
Err:1 http://archive.ubuntu.com/ubuntu focal InRelease
  Connection failed [IP: 10.0.3.4 8000]
Err:2 http://security.ubuntu.com/ubuntu focal-security InRelease
  Connection failed [IP: 10.0.3.4 8000]
```

**Comment by jemrobinson** [comment]
Can you attach logs from the proxy server too? Could be that these aren't supported repositories or that they're blocked by the firewall.

**Comment by craddm** [comment]
These are the exact same things that were failing earlier, so not clear why the firewall would be blocking them now; besides, at least one of them *sometimes* works:

```
shmblueadmin@SHM-BLUE-MANY-REPOSITORY-PROXY-TIER-2:~$ sudo apt-get update
Hit:1 http://archive.ubuntu.com/ubuntu focal InRelease
Hit:2 http://archive.ubuntu.com/ubuntu focal-updates InRelease
Hit:3 http://archive.ubuntu.com/ubuntu focal-backports InRelease
Err:4 http://security.ubuntu.com/ubuntu focal-security InRelease
  Connection failed [IP: 10.0.3.4 8000]
Reading package lists... Done
W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease  Connection failed [IP: 10.0.3.4 8000]
W: Some index files failed to download. They have been ignored, or old ones used instead.
```

Above, the initial `focal` repo is `Hit`, but on a subsequent run, it's `Err`:

```
shmblueadmin@SHM-BLUE-MANY-REPOSITORY-PROXY-TIER-2:~$ sudo apt-get update
Err:1 http://security.ubuntu.com/ubuntu focal-security InRelease
  Connection failed [IP: 10.0.3.4 8000]
Err:2 http://archive.ubuntu.com/ubuntu focal InRelease
  Connection failed [IP: 10.0.3.4 8000]
Hit:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease
Hit:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease
Reading package lists... Done
W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/focal/InRelease  Connection failed [IP: 10.0.3.4 8000]
W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease  Connection failed [IP: 10.0.3.4 8000]
W: Some index files failed to download. They have been ignored, or old ones used instead.
```

**Comment by craddm** [comment]
```
shmblueadmin@LINUX-UPDATES-SHM-BLUE:~$ sudo tail /var/log/squid-deb-proxy/access.log -n 25
1720537921.102     81 10.10.2.198 TCP_REFRESH_UNMODIFIED/200 105344 GET http://archive.ubuntu.com/ubuntu/dists/focal/multiverse/i18n/by-hash/SHA256/5d959b13455f2f7f040fdca1ed5d1a4d1860c4279096280fc621d408821d2c23 - HIER_DIRECT/91.189.91.82 application/octet-stream
1720537950.441  30030 10.10.2.198 TCP_REFRESH_ABORTED/000 0 GET http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease - HIER_DIRECT/185.125.190.81 -
1720537980.466  30023 10.10.2.198 TCP_REFRESH_ABORTED/000 0 GET http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease - HIER_DIRECT/185.125.190.81 -
1720538073.283     18 10.10.2.198 TCP_REFRESH_UNMODIFIED/304 447 GET http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease - HIER_DIRECT/91.189.91.82 -
1720538073.427      0 10.10.2.198 TCP_DENIED/403 3885 CONNECT esm.ubuntu.com:443 - HIER_NONE/- text/html
1720538073.431      0 10.10.2.198 TCP_DENIED/403 3885 CONNECT esm.ubuntu.com:443 - HIER_NONE/- text/html
1720538103.295  30030 10.10.2.198 TCP_MISS_ABORTED/000 0 GET http://archive.ubuntu.com/ubuntu/dists/focal/InRelease - HIER_DIRECT/185.125.190.81 -
1720538133.322  30025 10.10.2.198 TCP_MISS_ABORTED/000 0 GET http://archive.ubuntu.com/ubuntu/dists/focal/InRelease - HIER_DIRECT/185.125.190.81 -
1720538133.568    244 10.10.2.198 TCP_REFRESH_UNMODIFIED/304 447 GET http://archive.ubuntu.com/ubuntu/dists/focal-updates/InRelease - HIER_DIRECT/91.189.91.82 -
1720538133.649     79 10.10.2.198 TCP_REFRESH_UNMODIFIED/304 447 GET http://archive.ubuntu.com/ubuntu/dists/focal-backports/InRelease - HIER_DIRECT/91.189.91.82 -
1720538160.650      0 10.10.2.198 TCP_DENIED/403 3885 CONNECT esm.ubuntu.com:443 - HIER_NONE/- text/html
1720538160.653      0 10.10.2.198 TCP_DENIED/403 3885 CONNECT esm.ubuntu.com:443 - HIER_NONE/- text/html
1720538160.673    156 10.10.2.198 TCP_MISS/304 441 GET http://archive.ubuntu.com/ubuntu/dists/focal/InRelease - HIER_DIRECT/91.189.91.82 -
1720538160.753     77 10.10.2.198 TCP_REFRESH_UNMODIFIED/304 447 GET http://archive.ubuntu.com/ubuntu/dists/focal-updates/InRelease - HIER_DIRECT/91.189.91.82 -
1720538160.833     77 10.10.2.198 TCP_REFRESH_UNMODIFIED/304 447 GET http://archive.ubuntu.com/ubuntu/dists/focal-backports/InRelease - HIER_DIRECT/91.189.91.82 -
1720538190.540  30030 10.10.2.198 TCP_REFRESH_ABORTED/000 0 GET http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease - HIER_DIRECT/185.125.190.81 -
1720538220.574  30030 10.10.2.198 TCP_REFRESH_ABORTED/000 0 GET http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease - HIER_DIRECT/185.125.190.81 -
1720538336.647      0 10.10.2.198 TCP_DENIED/403 3885 CONNECT esm.ubuntu.com:443 - HIER_NONE/- text/html
1720538336.650      0 10.10.2.198 TCP_DENIED/403 3885 CONNECT esm.ubuntu.com:443 - HIER_NONE/- text/html
1720538366.538  30021 10.10.2.198 TCP_REFRESH_ABORTED/000 0 GET http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease - HIER_DIRECT/185.125.190.82 -
1720538366.546  30028 10.10.2.198 TCP_MISS_ABORTED/000 0 GET http://archive.ubuntu.com/ubuntu/dists/focal/InRelease - HIER_DIRECT/185.125.190.82 -
1720538396.571  30030 10.10.2.198 TCP_REFRESH_ABORTED/000 0 GET http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease - HIER_DIRECT/185.125.190.82 -
1720538396.578  30030 10.10.2.198 TCP_MISS_ABORTED/000 0 GET http://archive.ubuntu.com/ubuntu/dists/focal/InRelease - HIER_DIRECT/185.125.190.82 -
1720538396.750    170 10.10.2.198 TCP_REFRESH_UNMODIFIED/304 447 GET http://archive.ubuntu.com/ubuntu/dists/focal-updates/InRelease - HIER_DIRECT/91.189.91.82 -
1720538396.830     79 10.10.2.198 TCP_REFRESH_UNMODIFIED/304 447 GET http://archive.ubuntu.com/ubuntu/dists/focal-backports/InRelease - HIER_DIRECT/91.189.91.82 -
```

The only `denied` here is for `esm.ubuntu.com`

**Comment by craddm** [comment]
I think this was a re-emergence of the old problem with not having all the Ubuntu archive IP addresses explicitly allowed - there are at least 3 addresses missing from the below list, from `Configuration.psm1`.

```
        updateServers     = [ordered]@{
            externalIpAddresses = [ordered]@{
                linux   = (
                    @("72.32.157.246", "87.238.57.227", "147.75.85.69", "217.196.149.55") + # apt.postgresql.org
                    @("91.189.91.38", "91.189.91.39", "91.189.91.48", "91.189.91.49", "91.189.91.81", "91.189.91.82", "91.189.91.83", "185.125.190.17", "185.125.190.18", "185.125.190.36", "185.125.190.39") + # archive.ubuntu.com, changelogs.ubuntu.com, security.ubuntu.com
                    $cloudFlareIpAddresses + # database.clamav.net, packages.gitlab.com and qgis.org use Cloudflare
                    $cloudFrontIpAddresses + # packages.gitlab.com uses Cloudfront to host its Release file
                    @("104.131.190.124") + # dbeaver.io
                    @("152.199.20.126") + # developer.download.nvidia.com
                    $microsoftIpAddresses # packages.microsoft.com, azure.archive.ubuntu.com
```

**Comment by JimMadge** [comment]
@craddm Did we not backport the change to networking rules into v4?

Is the only work left here to replicate the fix across all containers?

**Comment by craddm** [comment]
> @craddm Did we not backport the change to networking rules into v4?
> 
> Is the only work left here to replicate the fix across all containers?

If you mean this [PR](https://github.com/alan-turing-institute/data-safe-haven/pull/1745/files), that didn't change anything for apt servers. 

Yes, I need to replicate the fix across all containers, but it's hard to be sure it's working, as the VMs are intermittently unable to download some software using apt

I'll get the fix working for all the required VMs, then open a separate issue for the apt problems

**Comment by craddm** [comment]
all working now. Last thing to do will be to document adding the docker credentials

**Comment by craddm** [comment]
What do we think is the best thing to do about Docker credentials with our example config files?

**Comment by jemrobinson** [comment]
> What do we think is the best thing to do about Docker credentials with our example config files?

Do we expect our example configs to work out-of-the-box? I assume not? If so, put some entries that are in the right format but are clearly dummy data.

**Comment by craddm** [comment]
> Do we expect our example configs to work out-of-the-box? I assume not? If so, put some entries that are in the right format but are clearly dummy data.

No, that's what I thought. 

One slight concern I have is that if things do wrong with the Docker part - e.g. no or incorrect credentials are supplied - it's going to fail silently in a non-obvious way.

**Comment by JimMadge** [comment]
> One slight concern I have is that if things do wrong with the Docker part - e.g. no or incorrect credentials are supplied - it's going to fail silently in a non-obvious way.

Could we run `docker login` locally to check?

**Comment by jemrobinson** [comment]
To be honest, things already go wrong with the Docker side of things all the time in very opaque ways (most often by being unable to install `docker.io`). I'd suggest this is a "will-not-fix" for a version of the code that we're going to stop supporting soon.

**Review by JimMadge — APPROVED** [review]

**Comment by craddm** [comment]
To be fair, I suspect a lot of the time it'll work *anyway* - if the login fails I think it'll just fall back to doing anonymous d/load

**Pull request merged**